mistral-rs: init at 0.1.18

GaetanLepage opened the pull request:
## Description of changes

Add [mistral-rs](https://github.com/EricLBuehler/mistral.rs), an LLM inference tool written in rust (similar to llama.cpp).

- [x] build the thing
- [ ] python package ?

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

GaetanLepage committed:
mistral-rs: init at 0.1.18

drupol reviewed (approved):
LGTM

GaetanLepage commented:
> LGTM

Well, it unfortunately does not build.

drupol commented:
Syntactically wise, it's fine. Now for the build, I'm pretty sure this is just a matter of time :)

GaetanLepage commented:
```
error[E0308]: mismatched types
   --> mistralrs-core/src/vision_models/phi3.rs:704:31
    |
704 | ...                   &[p_0..p_0 + 1, p_1..p_1 + cnt, 0..img_set_tensor.dims()[2]],
    |                         ^^^^^^^^^^^^ expected `&dyn RangeBound`, found `Range<usize>`
    |
    = note: expected reference `&dyn candle_core::tensor_indexing::RangeBound`
                  found struct `std::ops::Range<usize>`
help: consider borrowing here
    |
704 |                             &[&(p_0..p_0 + 1), p_1..p_1 + cnt, 0..img_set_tensor.dims()[2]],
    |                               ++            +

error[E0308]: mismatched types
   --> mistralrs-core/src/vision_models/phi3.rs:725:31
    |
725 | ...                   &[p_0..p_0 + 1, p_1..p_1 + cnt, 0..img_set_tensor.dims()[2]],
    |                         ^^^^^^^^^^^^ expected `&dyn RangeBound`, found `Range<usize>`
    |
    = note: expected reference `&dyn candle_core::tensor_indexing::RangeBound`
                  found struct `std::ops::Range<usize>`
help: consider borrowing here
    |
725 |                             &[&(p_0..p_0 + 1), p_1..p_1 + cnt, 0..img_set_tensor.dims()[2]],
    |                               ++            +

For more information about this error, try `rustc --explain E0308`.
error: could not compile `mistralrs-core` (lib) due to 2 previous errors

drupol commented:
2 things that comes to my mind:

1. Which Rust version does this project require
2. Have you checked the existing Dockerfile to find some pointers? https://github.com/EricLBuehler/mistral.rs/blob/master/Dockerfile

GaetanLepage commented:
@EricLBuehler, were you aware of such compilation errors ?
Could it be caused by the version of rust used here (1.77.1) ?

EricLBuehler commented:
@GaetanLepage glad to help! This is caused by an old version of our dependencies and not the Rust version. I just merged EricLBuehler/mistral.rs#419 (and also just released 0.1.17) which locks our Candle version, so any breaking changes do not affect mistral.rs.

GaetanLepage commented:
> @GaetanLepage glad to help! This is caused by an old version of our dependencies and not the Rust version. I just merged [EricLBuehler/mistral.rs#419](https://github.com/EricLBuehler/mistral.rs/pull/419) (and also just released 0.1.17) which locks our Candle version, so any breaking changes do not affect mistral.rs.

Thank you so much for the very quick solution. Everything seems to work now !

GaetanLepage commented:
Result of `nixpkgs-review pr 319008` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mistral-rs</li>
  </ul>
</details>

GaetanLepage commented:
Result of `nixpkgs-review pr 319008` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>mistral-rs</li>
  </ul>
</details>

GaetanLepage commented:
Result of `nixpkgs-review pr 319008` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>mistral-rs</li>
  </ul>
</details>

GaetanLepage commented:
Result of `nixpkgs-review pr 319008` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>mistral-rs</li>
  </ul>
</details>

GaetanLepage commented:
Result of `nixpkgs-review pr 319008` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>mistral-rs</li>
  </ul>
</details>

GaetanLepage commented:
Result of `nixpkgs-review pr 319008` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>mistral-rs</li>
  </ul>
</details>

drupol commented:
Status of this PR ?

GaetanLepage commented:
> Status of this PR ?

WIP: I have to implement the different acceleration schemes (cuda, metal, ...).

GaetanLepage commented on pkgs/by-name/mi/mistral-rs/package.nix:
```diff
@@ -0,0 +1,189 @@
+  cargoBuildFlags =
+    lib.optionals cudaAccel [ "--features=cuda" ]
+    ++ lib.optionals mklAccel [ "--features=mkl" ]
+    ++ lib.optionals stdenv.isDarwin [ "--features=metal" ];
```
Should we make `metal` optional on Darwin ? Like so, we could maybe build mistral-rs on `x86_64-darwin`...

SomeoneSerge replied:
Probably doesn't hurt to do `enableMetal ? isDarwin`

GaetanLepage replied:
Should I just get rid of the `accel` parameter and simply rely on `xxxxSupport` similarly to what [llama-cpp does](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/ll/llama-cpp/package.nix#L10) ?

SomeoneSerge replied:
Oh, I didn't realize this is just a variable from a let-in environment

GaetanLepage replied:
> Should I just get rid of the `accel` parameter and simply rely on `xxxxSupport` similarly to what [llama-cpp does](https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/ll/llama-cpp/package.nix#L10) ?

What is your opinion on this @SomeoneSerge ?

SomeoneSerge replied:
If mistral's support for backends is exclusive (only one allowed at a time) let's keep your current solution

GaetanLepage commented:
Current state:
- [x] Normal build -> OK for all platforms (except for `x86_64-darwin` because it doesn't have `darwin.apple_sdk.frameworks.MetalPerformanceShaders`)
- [x] `cuda` -> OK on both `x86_64-linux` and `aarch64-linux`
- [ ] `mkl` -> broken (complains about missing symbols)

TanvirOnGH commented:
[v0.1.23](https://github.com/EricLBuehler/mistral.rs/pull/495) is going to be released very soon

EricLBuehler commented:
@TanvirOnGH, I actually just [released v0.1.23](https://github.com/EricLBuehler/mistral.rs/releases/tag/v0.1.23)! Since version 0.1.18, we have added the Idefics 2 vision model and added the Gemma 2 model, as well as various optimizations, fixes, and improvements - but the APIs remain backwards compatible as per semver.

GaetanLepage commented:
> @TanvirOnGH, @GaetanLepage since version 0.1.18, we have added the Idefics 2 vision model and added the Gemma 2 model, as well as various optimizations, fixes, and improvements - but the APIs remain backwards compatible as per semver.

Thanks for the preview !
I am aware of the latest releases but didn't include them in this PR as we don't yet have Rust 1.79 in the `master` branch.
If my testing were correct, 0.1.18 is the latest mistral-rs version that can work with Rust 1.78.

EricLBuehler commented:
> > @TanvirOnGH, @GaetanLepage since version 0.1.18, we have added the Idefics 2 vision model and added the Gemma 2 model, as well as various optimizations, fixes, and improvements - but the APIs remain backwards compatible as per semver.
> 
> Thanks for the preview ! I am aware of the latest releases but didn't include them in this PR as we don't yet have Rust 1.79 in the `master` branch. If my testing were correct, 0.1.18 is the latest mistral-rs version that can work with Rust 1.78.

That makes sense. I just merged a PR to decrease our MSRV to 1.75.0, and we'll upload the new tag (it will be 0.1.25) in a day. I'll let you know here when we do that!

SomeoneSerge commented on pkgs/by-name/mi/mistral-rs/package.nix:
```diff
@@ -0,0 +1,189 @@
+      CUDA_COMPUTE_CAP = cudaCapability';
+
+      # Apparently, cudart is enough: No need to provide the entire cudaPackages.cudatoolkit derivation.
+      CUDA_TOOLKIT_ROOT_DIR = cudaPackages.cuda_cudart;
```
For compatibility with https://github.com/NixOS/nixpkgs/pull/323056, make it `detDev cuda_cudart` or `getLib cuda_cudart` (whichever works)

GaetanLepage replied:
Done

SomeoneSerge commented on pkgs/by-name/mi/mistral-rs/package.nix:
```diff
@@ -0,0 +1,189 @@
+    "--skip=sampler::tests::test_gumbel_speculative"
+  ];
+
+  passthru = {
```
A runtime test showing (in the simplest and minimal way) how to use the package would be nice as well. I'm thinking approximately along the lines of 

- https://github.com/NixOS/nixpkgs/blob/780006118670cccb7a516966822f84e7a00e75ad/pkgs/development/python-modules/torch/gpu-checks.nix#L16-L27
- https://gist.github.com/SomeoneSerge/8207f77c78e92e787d6d1114b01ae51c#file-llama-2-7b-nix-L108-L117

Not a blocker

SomeoneSerge commented on pkgs/by-name/mi/mistral-rs/package.nix:
```diff
@@ -0,0 +1,189 @@
+  };
+
+  meta = {
+    description = "Blazingly fast LLM inference";
```
Maybe let's just mark `mkl` as broken and merge

GaetanLepage replied:
Done

EricLBuehler commented on pkgs/by-name/mi/mistral-rs/package.nix:
```diff
@@ -0,0 +1,189 @@
+
+rustPlatform.buildRustPackage rec {
+  pname = "mistral-rs";
+  version = "0.1.18";
```
I think that this can be updated to 0.1.23?

GaetanLepage replied:
I will do this in a subsequent PR.

GaetanLepage commented:
Result of `nixpkgs-review pr 319008` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mistral-rs</li>
  </ul>
</details>

GaetanLepage commented:
Result of `nixpkgs-review pr 319008` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mistral-rs</li>
  </ul>
</details>

GaetanLepage commented:
Result of `nixpkgs-review pr 319008` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)

GaetanLepage commented:
Result of `nixpkgs-review pr 319008` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)

GaetanLepage commented:
Result of `nixpkgs-review pr 319008` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mistral-rs</li>
  </ul>
</details>

GaetanLepage commented:
Result of `nixpkgs-review pr 319008` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mistral-rs</li>
  </ul>
</details>

GaetanLepage commented:
Result of `nixpkgs-review pr 319008` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mistral-rs</li>
  </ul>
</details>

GaetanLepage commented:
Result of `nixpkgs-review pr 319008` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>mistral-rs</li>
  </ul>
</details>

SomeoneSerge commented on pkgs/by-name/mi/mistral-rs/package.nix:
```diff
@@ -0,0 +1,201 @@
+  inherit (cudaPackages.cudaFlags) cudaCapabilities;
+  cudaCapabilityString =
+    if cudaCapability == null then
+      (builtins.head (
+        (builtins.filter (cap: lib.versionAtLeast cap minRequiredCudaCapability) cudaCapabilities)
```
@EricLBuehler it's very convenient to have you here 🙃 Do you think it could be possible to pass a _list_ of gencodes through mistral-rs' build rules to nvcc, instead of just a single architecture? In Nixpkgs the user configures the expected cuda architectures globally as a list, but @GaetanLepage's current solution is to pick the oldest of the requested architectures and hope that the PTX code is included in the output.

GaetanLepage replied:
The compute capability is used by the `bindgen_cuda` crate. It gives its value [to the `--architecture` flag](https://github.com/Narsil/bindgen_cuda/blob/a6b0c891be8ebefb55600d46d8a77a358adc3913/src/lib.rs#L246).

SomeoneSerge replied:
@GaetanLepage Out of scope for the PR, but we could try burning the gencode flags into `NVCC_APPEND_FLAGS`, i.e. we'd tell nvcc to build fatbins without the rust component even noticiting

EricLBuehler replied:
Yeah, I think that's a good idea. Mistral.rs doesn't directly call nvcc as we delegate it to `bindgen_cuda` as @GaetanLepage pointed out. The best way is probably to use some environment variable.

GaetanLepage commented:
@SomeoneSerge @EricLBuehler Thank you so much for your help.
Should we merge this as is ?

SomeoneSerge commented on pkgs/by-name/mi/mistral-rs/package.nix:
```diff
@@ -0,0 +1,201 @@
+          #   https://github.com/juhaku/utoipa/blob/utoipa-swagger-ui-<UTOPIA-SWAGGER-UI-VERSION>/utoipa-swagger-ui/build.rs#L21-L22
+          swaggerUiVersion = "5.17.12";
+
+          swaggerUi = fetchurl {
+            url = "https://github.com/swagger-api/swagger-ui/archive/refs/tags/v${swaggerUiVersion}.zip";
+            hash = "sha256-HK4z/JI+1yq8BTBJveYXv9bpN/sXru7bn/8g5mf2B/I=";
+          };
+        in
+        "file://${swaggerUi}";
```
I see there's something called `swagger-ui-bundle` in `python3Packages` but no source build. Can we eventually migrate the mistral-rs package to use a top-level attribute instead of the ad hoc fetchurl?..

EricLBuehler commented:
@GaetanLepage, @SomeoneSerge thank you!
